contrib/platform/src/com/sun/jna/platform/win32/Wtsapi32.java

```diff
@@ -113,6 +113,100 @@
     int WTS_PROCESS_INFO_LEVEL_0 = 0;
     int WTS_PROCESS_INFO_LEVEL_1 = 1;
 
+    int DOMAIN_LENGTH = 17;
+    int USERNAME_LENGTH = 20;
+    int WINSTATIONNAME_LENGTH = 32;
+
+    /**
+     * Contains information about a client session on a Remote Desktop Session Host
+     * (RD Session Host) server.
+     */
+    @FieldOrder({ "SessionId", "pWinStationName", "State" })
+    class WTS_SESSION_INFO extends Structure {
+        public int SessionId;
+        public String pWinStationName; // either LPSTR or LPWSTR
+        public int State; // WTS_CONNECTSTATE_CLASS
+
+        public WTS_SESSION_INFO() {
+            super(W32APITypeMapper.DEFAULT);
+        }
+
+        public WTS_SESSION_INFO(Pointer p) {
+            super(p, Structure.ALIGN_DEFAULT, W32APITypeMapper.DEFAULT);
+            read();
+        }
+    }
+
+    /**
+     * Contains the client network address of a Remote Desktop Services session.
+     */
+    @FieldOrder({ "AddressFamily", "Address" })
+    class WTS_CLIENT_ADDRESS extends Structure {
+        public int AddressFamily;
+        public byte[] Address = new byte[20];
+
+        public WTS_CLIENT_ADDRESS() {
+            super();
+        }
+
+        public WTS_CLIENT_ADDRESS(Pointer p) {
+            super(p);
+            read();
+        }
+    }
+
+    /**
+     * Contains information about a Remote Desktop Services session.
+     */
+    @FieldOrder({ "State", "SessionId", "IncomingBytes", "OutgoingBytes", "IncomingFrames", "OutgoingFrames",
+            "IncomingCompressedBytes", "OutgoingCompressedBytes", "WinStationName", "Domain", "UserName", "ConnectTime",
+            "DisconnectTime", "LastInputTime", "LogonTime", "CurrentTime" })
+    class WTSINFO extends Structure {
+        private static final int CHAR_WIDTH = Boolean.getBoolean("w32.ascii") ? 1 : 2;
+
+        public int State;
+        public int SessionId;
+        public int IncomingBytes;
+        public int OutgoingBytes;
+        public int IncomingFrames;
+        public int OutgoingFrames;
+        public int IncomingCompressedBytes;
+        public int OutgoingCompressedBytes;
+        public byte[] WinStationName = new byte[WINSTATIONNAME_LENGTH * CHAR_WIDTH];
+        public byte[] Domain = new byte[DOMAIN_LENGTH * CHAR_WIDTH];
+        public byte[] UserName = new byte[(USERNAME_LENGTH + 1) * CHAR_WIDTH];
```

Comment:
I would add a javadoc comment to the three parameters, that they exist only to enable structure size calculation. I would also make the three fields `final` (see https://java-native-access.github.io/jna/5.5.0/javadoc/overview-summary.html -> "read-only" fields).

Reading from native memory and will still work. Even a setter should work that way.

Reply:
Regarding `final` that requires the fields be initialized in the constructor.  When constructing from a `Pointer` this works.  Should I then remove the no-arg constructor?

Reply:
Actually... as arrays, final doesn't really change writing to the contents. So this doesn't prevent changing the contents but does fix the length, and thus the structure offsets. Not really "read-only" but still a good idea, probably JNA-wide for arrays in structures...